File: backend/services/routing.py
import json
import networkx as nx
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import text
from schemas import RouteRequest

async def find_safe_route(session: AsyncSession, req: RouteRequest) -> dict:
    # Шаг А: Поиск ближайших узлов старта и финиша в графе
    closest_node_query = text("""
        SELECT id
        FROM nodes
        ORDER BY geom <-> ST_SetSRID(ST_MakePoint(:lon, :lat), 4326)
        LIMIT 1;
    """)
    
    start_res = await session.execute(closest_node_query, {"lon": req.start_lon, "lat": req.start_lat})
    start_node_id = start_res.scalar_one()

    end_res = await session.execute(closest_node_query, {"lon": req.end_lon, "lat": req.end_lat})
    end_node_id = end_res.scalar_one()

    # Шаг Б: Загрузка всех ребер для построения in-memory графа
    edges_query = text("""
        SELECT 
            id,
            start_node_id, 
            end_node_id, 
            base_weight, 
            current_risk_weight, 
            ST_AsGeoJSON(geom) as geometry
        FROM edges;
    """)
    
    edges_res = await session.execute(edges_query)
    edges = edges_res.fetchall()

    # Шаг В: Билд графа через NetworkX
    G = nx.Graph()
    edge_geometries = {}

    for edge in edges:
        # Умный комбинированный вес = длина дороги + штраф за уровень угрозы!
        total_weight = edge.base_weight + edge.current_risk_weight
        
        # Добавляем ребро в граф
        G.add_edge(edge.start_node_id, edge.end_node_id, weight=total_weight)
        
        # Кешируем GeoJSON ребра, чтобы моментально собрать геометрию пути (связь двунаправленная)
        edge_geometries[(edge.start_node_id, edge.end_node_id)] = edge.geometry
        edge_geometries[(edge.end_node_id, edge.start_node_id)] = edge.geometry

    # Шаг Г: Алгоритм Дейкстры (или A*) на минимальный суммарный "штрафной" вес
    try:
        path_nodes = nx.shortest_path(G, source=start_node_id, target=end_node_id, weight='weight')
    except nx.NetworkXNoPath:
        return {"type": "FeatureCollection", "features": []}

    # Шаг Д: Сборка финального GeoJSON для отдачи на фронтенд
    features = []
    
    for i in range(len(path_nodes) - 1):
        u = path_nodes[i]
        v = path_nodes[i+1]
        
        geom_json_str = edge_geometries.get((u, v))
        if geom_json_str:
            features.append({
                "type": "Feature",
                "properties": {
                    "is_safe_route": True
                },
                "geometry": json.loads(geom_json_str)
            })

    return {
        "type": "FeatureCollection",
        "features": features
    }

File: rust_backend/src/routing.rs
use axum::{
    extract::State,
    Json,
};
use serde_json::{json, Value};
use sqlx::Row;
use petgraph::graphmap::UnGraphMap;
use petgraph::algo::astar;
use std::sync::Arc;
use std::collections::HashMap;

use crate::{AppState, models::RouteRequest};

pub async fn calculate_route(
    State(state): State<Arc<AppState>>,
    Json(req): Json<RouteRequest>,
) -> Json<Value> {
    tracing::info!("Поиск безопасного маршрута между ({}, {}) и ({}, {})", req.start_lon, req.start_lat, req.end_lon, req.end_lat);

    // 1. Поиск ближайших узлов старта и финиша в графе
    let node_query = r#"
        SELECT id
        FROM nodes
        ORDER BY geom <-> ST_SetSRID(ST_MakePoint($1, $2), 4326)
        LIMIT 1;
    "#;

    let start_node: i64 = match sqlx::query(node_query)
        .bind(req.start_lon)
        .bind(req.start_lat)
        .fetch_one(&state.db).await {
            Ok(row) => row.get("id"),
            Err(e) => {
                tracing::error!("Start node not found: {}", e);
                return Json(json!({"type": "FeatureCollection", "features": []}));
            }
        };

    let end_node: i64 = match sqlx::query(node_query)
        .bind(req.end_lon)
        .bind(req.end_lat)
        .fetch_one(&state.db).await {
            Ok(row) => row.get("id"),
            Err(e) => {
                tracing::error!("End node not found: {}", e);
                return Json(json!({"type": "FeatureCollection", "features": []}));
            }
        };

    // 2. Получение всех ребер для построения in-memory графа (Аналог nx.Graph())
    let edges_query = r#"
        SELECT 
            start_node_id, 
            end_node_id, 
            base_weight, 
            current_risk_weight, 
            ST_AsGeoJSON(geom) as geometry
        FROM edges;
    "#;

    let rows = match sqlx::query(edges_query).fetch_all(&state.db).await {
        Ok(res) => res,
        Err(e) => {
            tracing::error!("Error fetching edges: {}", e);
            return Json(json!({"type": "FeatureCollection", "features": []}));
        }
    };

    // Строим неориентированный граф (UnGraphMap)
    let mut graph = UnGraphMap::<i64, f64>::with_capacity(rows.len(), rows.len() * 2);
    let mut geometries = HashMap::with_capacity(rows.len());

    for row in rows {
        let u: i64 = row.get("start_node_id");
        let v: i64 = row.get("end_node_id");
        let base_weight: f64 = row.get("base_weight");
        let current_risk_weight: f64 = row.get("current_risk_weight");
        
        // Ключевая формула безопасного пути (как в Питоне)
        let weight = base_weight + current_risk_weight;
        graph.add_edge(u, v, weight);

        if let Ok(geom_str) = row.try_get::<String, _>("geometry") {
            // Геометрия не зависит от направления ребра
            let key = if u < v { (u, v) } else { (v, u) };
            geometries.insert(key, geom_str);
        }
    }

    // 3. Вычисление кратчайшего пути (A*)
    tracing::info!("Билд графа завершен. Запуск A*...");
    let path = astar(
        &graph,
        start_node,
        |finish| finish == end_node,
        |e| *e.2, // e - это кортеж (откуда, куда, ссылка_на_вес)
        |_| 0.0, // эвристика равна нулю = алгоритм Дейкстры
    );

    // 4. Сборка GeoJSON для фронтенда
    let mut features = Vec::new();

    if let Some((total_weight, nodes)) = path {
        tracing::info!("Маршрут найден! Вес: {}, Узлов: {}", total_weight, nodes.len());
        for i in 0..nodes.len() - 1 {
            let u = nodes[i];
            let v = nodes[i+1];
            let key = if u < v { (u, v) } else { (v, u) };

            if let Some(geom_str) = geometries.get(&key) {
                let geom_json: Value = serde_json::from_str(geom_str).unwrap_or(json!({}));
                features.push(json!({
                    "type": "Feature",
                    "properties": {
                        "is_safe_route": true
                    },
                    "geometry": geom_json
                }));
            }
        }
    } else {
        tracing::warn!("Маршрут не найден!");
    }

    Json(json!({
        "type": "FeatureCollection",
        "features": features
    }))
}
